### PR TITLE
Handle SIGINT and hook up service and adapter "stop()" methods.

### PR DESCRIPTION
**Pull request by mcav**
r? @fabricedesre

### CODE REVIEW

File: src/adapters/mod.rs [hunk 1]
```diff
@@ -5,20 +5,38 @@
...
 
 use controller::Controller;
+use self::philips_hue::PhilipsHueAdapter;
 use service::ServiceAdapter;
+use std::sync::{ Arc, Mutex };
 
 pub struct AdapterManager<T> {
-    controller: T
+    controller: T,
+    adapters: Arc<Mutex<Vec<Box<ServiceAdapter>>>>
```
Comment (fabricedesre) [inline src/adapters/mod.rs]: Why do we need an Arc<Mutex<>> here? AdapterManager is always used on the same thread.

File: src/adapters/mod.rs [hunk 2]
```diff
@@ -5,20 +5,38 @@
...
+    fn start_adapter(&self, adapter: Box<ServiceAdapter>) {
+        adapter.start();
+        self.adapters.lock().unwrap().push(adapter);
+    }
+
+    /// Stop all the adapters.
+    pub fn stop(&self) {
+        for adapter in &*self.adapters.lock().unwrap() {
+            adapter.stop();
+        }
```
Comment (fabricedesre) [inline src/adapters/mod.rs:40]: nit: Indentation is off because you have tabs in this file

File: src/main.rs [hunk 3]
```diff
@@ -117,7 +120,19 @@ fn update_hostname(hostname: String) -> Option<String> {
     Some(host_manager.set_name(&hostname))
 }
 
+// Handle SIGINT (Ctrl-C) for manual shutdown.
+// Signal handlers must not do anything substantial. To trigger shutdown, we atomically
+// flip this flag; the event loop checks the flag and exits accordingly.
+static INTERRUPTED_FLAG: AtomicBool = ATOMIC_BOOL_INIT;
```
Comment (fabricedesre) [inline src/main.rs]: nit: s/INTERRUPTED_FLAG/SHUTDOWN_FLAG

**Comment by fabricedesre** [comment]
One missing piece is to stop the tunnel's managed process. But it's fine to do that in a follow up.

**Comment by mcav** [comment]
@fabricedesre I've updated the PR to address your comments, and a second commit to stop the tunnel on shutdown.

File: src/controller.rs [hunk 4]
```diff
@@ -38,7 +39,7 @@ const DEFAULT_HOSTNAME: &'static str = "::"; // ipv6 default.
 const DEFAULT_DOMAIN: &'static str = ".local";
 
 pub trait Controller : Send + Sync + Clone + Reflect + 'static {
-    fn run(&mut self);
+    fn run(&mut self, interrupted_flag: &AtomicBool);
```
Comment (fabricedesre) [inline src/controller.rs]: nit: rename all the interrupted_flag to shutdown_flag

**Comment by fabricedesre** [comment]
Thanks! Feel free to merge once the small nits are fixed.

File: src/controller.rs (hunk 4 above)
Comment (mcav) [inline src/controller.rs]: ah, case-sensitive search strikes again!

**Pull request merged**